lib/chatbot-api/appsync-ws.ts

```diff
@@ -0,0 +1,133 @@
+import * as cdk from "aws-cdk-lib";
+import * as appsync from "aws-cdk-lib/aws-appsync";
+import { Code, Function, LayerVersion, Runtime } from "aws-cdk-lib/aws-lambda";
+import { SqsEventSource } from "aws-cdk-lib/aws-lambda-event-sources";
+import { Construct } from "constructs";
+import { Shared } from "../shared";
+import { IQueue } from "aws-cdk-lib/aws-sqs";
+import { ITopic } from "aws-cdk-lib/aws-sns";
+import { UserPool } from "aws-cdk-lib/aws-cognito";
+import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
+import * as path from "path";
+
+interface ChatGraphqlApiProps {
+  readonly queue: IQueue;
+  readonly topic: ITopic;
+  readonly userPool: UserPool;
+  readonly shared: Shared;
+}
+
+export class ChatGraphqlApi extends Construct {
+  public readonly apiKey: string | undefined;
+  public readonly graphQLUrl: string | undefined;
+  public readonly graphQLApi: appsync.GraphqlApi;
+  public readonly outgoingMessageHandler: Function;
+
+  constructor(scope: Construct, id: string, props: ChatGraphqlApiProps) {
+    super(scope, id);
+
+    const powertoolsLayerJS = LayerVersion.fromLayerVersionArn(
+      this,
+      "PowertoolsLayerJS",
+      `arn:aws:lambda:${
+        cdk.Stack.of(this).region
+      }:094274105915:layer:AWSLambdaPowertoolsTypeScript:22`
+    );
+
+    // makes a GraphQL API
+    const api = new appsync.GraphqlApi(this, "ws-api", {
+      name: "chatbot-ws-api",
+      schema: appsync.SchemaFile.fromAsset(
+        "lib/chatbot-api/schema/schema-ws.graphql"
+      ),
+      authorizationConfig: {
+        additionalAuthorizationModes: [
+          {
+            authorizationType: appsync.AuthorizationType.IAM,
+          },
+          {
+            authorizationType: appsync.AuthorizationType.USER_POOL,
+            userPoolConfig: {
+              userPool: props.userPool,
+            },
+          },
+        ],
+      },
+      xrayEnabled: true,
+    });
+
+    const resolverFunction = new Function(this, "lambda-resolver", {
+      code: Code.fromAsset(
+        "./lib/chatbot-api/functions/resolvers/send-query-lambda-resolver"
+      ),
+      handler: "index.handler",
+      runtime: Runtime.PYTHON_3_11,
+      environment: {
+        SNS_TOPIC_ARN: props.topic.topicArn,
+      },
+      layers: [props.shared.powerToolsLayer],
+    });
+
+    const outgoingMessageHandler = new NodejsFunction(
+      this,
+      "outgoing-message-handler",
+      {
+        entry: path.join(
+          __dirname,
+          "functions/outgoing-message-appsync/index.ts"
+        ),
+        layers: [powertoolsLayerJS],
+        handler: "index.handler",
+        runtime: Runtime.NODEJS_18_X,
+        environment: {
+          GRAPHQL_ENDPOINT: api.graphqlUrl,
+        },
+      }
+    );
```

Comment:
To prevent issues with Lambda Insights version mismatch, it's recommended to programmatically determine the Lambda Insights version by leveraging the CDK Variable as described here: https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Lambda-Insights-extension-versions.html

```suggestion
    // makes a GraphQL API
    const api = new appsync.GraphqlApi(this, "ws-api", {
      name: "chatbot-ws-api",
      schema: appsync.SchemaFile.fromAsset(
        "lib/chatbot-api/schema/schema-ws.graphql"
      ),
      authorizationConfig: {
        additionalAuthorizationModes: [
          {
            authorizationType: appsync.AuthorizationType.IAM,
          },
          {
            authorizationType: appsync.AuthorizationType.USER_POOL,
            userPoolConfig: {
              userPool: props.userPool,
            },
          },
        ],
      },
      xrayEnabled: true,
    });

    const resolverFunction = new Function(this, "lambda-resolver", {
      code: Code.fromAsset(
        "./lib/chatbot-api/functions/resolvers/send-query-lambda-resolver"
      ),
      handler: "index.handler",
      runtime: Runtime.PYTHON_3_11,
      environment: {
        SNS_TOPIC_ARN: props.topic.topicArn,
      },
      insightsVersion: LambdaInsightsVersion.VERSION_1_0_229_0,
    });

    const outgoingMessageHandler = new NodejsFunction(
      this,
      "outgoing-message-handler",
      {
        entry: path.join(
          __dirname,
          "functions/outgoing-message-appsync/index.ts"
        ),
        handler: "index.handler",
        runtime: Runtime.NODEJS_18_X,
        insightsVersion: LambdaInsightsVersion.VERSION_1_0_229_0,
        environment: {
          GRAPHQL_ENDPOINT: api.graphqlUrl,
        },
      }
    );
```

Reply:
We are not using Lambda Insights but Lambda Powertools.